fbpic/particles/gathering/cuda_methods.py

```diff
@@ -192,6 +199,169 @@ def gather_field_gpu_linear(x, y, z,
+        (is modified by this function)
+
+    averaging : boolean, optional
+        Whether to average the new values with the old ones or to
+        discard the old values.
+    """
+    # Get the 1D CUDA grid
+    i = cuda.grid(1)
+    # Deposit the field per cell in parallel
+    # (for threads < number of particles)
+    if i < x.shape[0]:
+        # Preliminary arrays for the cylindrical conversion
+        # --------------------------------------------
+        # Position
+        xj = x[i]
+        yj = y[i]
+        zj = z[i]
+
+        # Cylindrical conversion
+        rj = math.sqrt( xj**2 + yj**2 )
+        if (rj !=0. ) :
+            invr = 1./rj
+            cos = xj*invr  # Cosine
+            sin = yj*invr  # Sine
+        else :
+            cos = 1.
+            sin = 0.
+
+
+        # Get linear weights for the deposition
+        # --------------------------------------------
+        # Positions of the particles, in the cell unit
+        r_cell =  invdr*(rj - rmin) - 0.5
+        z_cell =  invdz*(zj - zmin) - 0.5
+        # Original index of the uppper and lower cell
+        ir_lower = int(math.floor( r_cell ))
+        ir_upper = ir_lower + 1
+        iz_lower = int(math.floor( z_cell ))
+        iz_upper = iz_lower + 1
+        # Linear weight
+        Sr_lower = ir_upper - r_cell
+        Sr_upper = r_cell - ir_lower
+        Sz_lower = iz_upper - z_cell
+        Sz_upper = z_cell - iz_lower
+        # Set guard weights to zero
+        Sr_guard = 0.
+
+        # Treat the boundary conditions
+        # --------------------------------------------
+        # guard cells in lower r
+        if ir_lower < 0:
+            Sr_guard = Sr_lower
+            Sr_lower = 0.
+            ir_lower = 0
+        # absorbing in upper r
+        if ir_lower > Nr-1:
+            ir_lower = Nr-1
+        if ir_upper > Nr-1:
+            ir_upper = Nr-1
+        # periodic boundaries in z
+        # lower z boundaries
+        if iz_lower < 0:
+            iz_lower += Nz
+        if iz_upper < 0:
+            iz_upper += Nz
+        # upper z boundaries
+        if iz_lower > Nz-1:
+            iz_lower -= Nz
+        if iz_upper > Nz-1:
+            iz_upper -= Nz
+
+        # Precalculate Shapes
+        S_ll = Sz_lower*Sr_lower
+        S_lu = Sz_lower*Sr_upper
+        S_ul = Sz_upper*Sr_lower
+        S_uu = Sz_upper*Sr_upper
+        S_lg = Sz_lower*Sr_guard
+        S_ug = Sz_upper*Sr_guard
+
+        # Envelope fields
+        # -------
+        F = 0
+        Fr = 0.
+        Ft = 0.
+        Fz = 0.
+
+        for m in m_tuple:
+            # Add contribution from mode m
+            a_m = a_tuple[m]
+            grad_a_r_m = grad_a_r_tuple[m]
+            grad_a_t_m = grad_a_t_tuple[m]
+            grad_a_z_m = grad_a_z_tuple[m]
+            # Calculate azimuthal complex factor
+            exptheta_m = 1.
+            for _ in range(m):
+                exptheta_m *= (cos - 1.j*sin)
```

Comment:
It seems that the above would not work for negative `m`. (This only impacts the GPU implementation: the CPU implementation uses a different expression.)

I would suggest replacing it with:
```
for _ in range(abs(m)):
    exptheta_m *= (cos - 1.j*sin)
if m < 0:
   exptheta_m = exp_theta.conjugate()
```
(This is a trick to avoid negative powers - i.e. divisions - which are more time-consuming than multiplications.)